## Case Técnico

In [2]:
# Bibliotecas para tratar dados

import pandas as pd
import numpy as np

# Bibliotecas para visualização de dados
import matplotlib as mpl
import matplotlib.pyplot as plt

# Configurar para não exibir warnings
import warnings
warnings.filterwarnings("ignore")

# Configurar para exibir até 15 linhas de um Dataframe do Pandas
pd.set_option("display.max_rows", 15)

# Configurar para exibir todas as colunas de um Dataframe do Pandas
pd.set_option("display.max_columns", None)


In [3]:
speech = pd.read_excel('BASE_SPEECH_CASE.xlsx')

Vamos exibir as primeiras 5 linhas da base para verificar como ela está estruturada.

In [4]:
speech.head()

audio_length        callId                    callDate  \
0   405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
1   405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
2   405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
3   405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
4   405.291992  1.546791e+11  2024-07-01 11:32:00:000000   

                  initiatedAt   segment                terminatedAt  \
0  2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
1  2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
2  2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
3  2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
4  2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   

                                reason reason_type  \
0                         Cancelamento      speech   
1              Script de Finalização      script   
2  Transferir Pesquisa de Satisfação      script   
3                             Bloqueio      speech   
4              Script de Finalização      script   

                              subreasons_description  position sgl_uf  \
0                             Solicitar cancelamento       2.0     SP   
1                            Script de Finalização      23.0     SP   
2  Transferir ligação para pesquisa de satisfaÃ...      21.0     SP   
3                   Informar bloqueio de linha/plano      18.0     SP   
4                            Script de Finalização      24.0     SP   

   customer    site plan_name  
0       1.0  Site 1   Plano 1  
1       1.0  Site 1   Plano 1  
2       1.0  Site 1   Plano 1  
3       1.0  Site 1   Plano 1  
4       1.0  Site 1   Plano 1

Podemos perceber que os dados acentuados estão com problema de codificação. Vamos começar a etapa de tratamento dos dados

## 1- Tratamento dos dados

Para resolver o problema de codificação dos dados, irei criar uma função chamada corrigir_codificacao e aplicarei-a na nossa base.

In [5]:
def corrigir_codificacao(base):
    if isinstance(base, str):
        try:
            return base.encode('latin1', errors='ignore').decode('utf-8', errors='ignore')
        except Exception:
            return base
    return base

speech = speech.map(corrigir_codificacao)

In [6]:
speech = speech.map(corrigir_codificacao)

In [7]:
speech

audio_length        callId                    callDate  \
0        405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
1        405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
2        405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
3        405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
4        405.291992  1.546791e+11  2024-07-01 11:32:00:000000   
...             ...           ...                         ...   
191069  1392.302979  1.546933e+11  2024-07-15 15:34:57:000000   
191070  1392.302979  1.546933e+11  2024-07-15 15:34:57:000000   
191071  1392.302979  1.546933e+11  2024-07-15 15:34:57:000000   
191072  1392.302979  1.546933e+11  2024-07-15 15:34:57:000000   
191073  1392.302979  1.546933e+11  2024-07-15 15:34:57:000000   

                       initiatedAt   segment                terminatedAt  \
0       2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
1       2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
2       2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
3       2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
4       2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
...                            ...       ...                         ...   
191069  2024-07-15 15:34:57:000000  pos puro  2024-07-15 15:58:09:303000   
191070  2024-07-15 15:34:57:000000  pos puro  2024-07-15 15:58:09:303000   
191071  2024-07-15 15:34:57:000000  pos puro  2024-07-15 15:58:09:303000   
191072  2024-07-15 15:34:57:000000  pos puro  2024-07-15 15:58:09:303000   
191073  2024-07-15 15:34:57:000000  pos puro  2024-07-15 15:58:09:303000   

                                          reason reason_type  \
0                                   Cancelamento      speech   
1                            Script de Finalizao      script   
2                Transferir Pesquisa de Satisfao      script   
3                                       Bloqueio      speech   
4                            Script de Finalizao      script   
...                                          ...         ...   
191069                      Informao sobre Plano      speech   
191070                     Script de Inicializao      script   
191071                           Confirmao Dados      script   
191072                        Informar Protocolo      script   
191073  Aplicativo: Instruo de uso do aplicativo      script   

                            subreasons_description  position sgl_uf  customer  \
0                           Solicitar cancelamento       2.0     SP       1.0   
1                              Script de Finalizao      23.0     SP       1.0   
2       Transferir ligao para pesquisa de satisfao      21.0     SP       1.0   
3                 Informar bloqueio de linha/plano      18.0     SP       1.0   
4                              Script de Finalizao      24.0     SP       1.0   
...                                            ...       ...    ...       ...   
191069                      Saber plano contratado      23.0     DF    5831.0   
191070                       Script de Inicializao      17.0     DF    5831.0   
191071            Confirmao de Nome, CPF, Telefone      12.0     DF    5831.0   
191072                          Informar Protocolo      19.0     DF    5831.0   
191073  Instruo de uso de aplicativos corporativos      10.0     DF    5831.0   

          site plan_name  
0       Site 1   Plano 1  
1       Site 1   Plano 1  
2       Site 1   Plano 1  
3       Site 1   Plano 1  
4       Site 1   Plano 1  
...        ...       ...  
191069  Site 4  Plano 50  
191070  Site 4  Plano 50  
191071  Site 4  Plano 50  
191072  Site 4  Plano 50  
191073  Site 4  Plano 50  

[191074 rows x 14 columns]

Com base nos dados acima, podemos perceber que o problema de codificação foi corrigido com sucesso! Vamos prosseguir com o tratamento de dados, focando agora em verificar se existem valores nulos em nossa base.

In [8]:
speech.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 191074 entries, 0 to 191073
Data columns (total 14 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   audio_length            191072 non-null  object 
 1   callId                  191072 non-null  float64
 2   callDate                191071 non-null  object 
 3   initiatedAt             191068 non-null  object 
 4   segment                 191072 non-null  object 
 5   terminatedAt            191072 non-null  object 
 6   reason                  191072 non-null  object 
 7   reason_type             191072 non-null  object 
 8   subreasons_description  191072 non-null  object 
 9   position                191072 non-null  float64
 10  sgl_uf                  191072 non-null  object 
 11  customer                191072 non-null  float64
 12  site                    191072 non-null  object 
 13  plan_name               191072 non-null  object 
dtypes: float64(3), objec

A base possui 191074 linhas. Porém, a maioria das colunas possui 191072 valores, sendo que a coluna **callDate** possui 191071 valores e coluna **initiatedAt** possui 191068 valores. Vamos investigar isso através da verificação dos valores nulos presentes na base.

In [9]:
speech[speech.isnull().any(axis=1)]

audio_length        callId                    callDate  \
587    955.469971  1.546902e+11                         NaN   
692    358.656006  1.546848e+10  2024-07-06 10:21:54:000000   
695    358.656006  1.546848e+10  2024-07-06 10:21:54:000000   
1123         1083  1.546855e+10  2024-07-07 10:36:36:000000   
1182    391.48999  1.546824e+11  2024-07-04 18:27:31:000000   
1326          NaN           NaN                         NaN   
1328          NaN           NaN                         NaN   

                     initiatedAt   segment                terminatedAt  \
587   2024-07-12 14:19:12:000000  pos puro  2024-07-12 14:35:07:470000   
692                          NaN  pos puro  2024-07-06 10:27:52:656000   
695                          NaN  pos puro  2024-07-06 10:27:52:656000   
1123                         NaN  pos puro  2024-07-07 10:54:39:000000   
1182                         NaN  pos puro  2024-07-04 18:34:02:490000   
1326                         NaN       NaN                         NaN   
1328                         NaN       NaN                         NaN   

                               reason reason_type  \
587              Informao sobre Plano      speech   
692             Script de Inicializao      script   
695              Questionar Algo Mais      script   
1123  Transferir Pesquisa de Satisfao      script   
1182         2a Via / Cdigo de Barras      speech   
1326                              NaN         NaN   
1328                              NaN         NaN   

                          subreasons_description  position sgl_uf  customer  \
587         Saber valor pago no plano contratado      10.0     SP     475.0   
692                        Script de Inicializao       0.0     AL     625.0   
695                         Questionar Algo Mais      11.0     AL     625.0   
1123  Transferir ligao para pesquisa de satisfao      10.0     SP     879.0   
1182             Solicitar segunda via de fatura       4.0     DF     896.0   
1326                                         NaN       NaN    NaN       NaN   
1328                                         NaN       NaN    NaN       NaN   

        site plan_name  
587   Site 1   Plano 1  
692   Site 1   Plano 1  
695   Site 1   Plano 1  
1123  Site 1   Plano 1  
1182  Site 1   Plano 1  
1326     NaN       NaN  
1328     NaN       NaN

De acordo com as informações acima, podemos perceber que existe 1 valor nulo na coluna **callDate**, 4 valores nulos na coluna **initiatedAt** e 2 linhas vazias. Primeiro vamos eliminar essas linhas vazias e depois trataremos os valores nulos.

In [10]:
speech = speech[speech.isnull().all(axis=1) == False]

Para tratar os valores nulos presentes nas colunas **callDate** e **initiatedAt** precisamos entender os padrões dessas colunas. Vou observar os valores presentes nas 5 primeiras e 5 ultimas linhas da base.

In [11]:
speech[['callDate', 'initiatedAt']]

callDate                 initiatedAt
0       2024-07-01 11:32:00:000000  2024-07-01 11:32:00:000000
1       2024-07-01 11:32:00:000000  2024-07-01 11:32:00:000000
2       2024-07-01 11:32:00:000000  2024-07-01 11:32:00:000000
3       2024-07-01 11:32:00:000000  2024-07-01 11:32:00:000000
4       2024-07-01 11:32:00:000000  2024-07-01 11:32:00:000000
...                            ...                         ...
191069  2024-07-15 15:34:57:000000  2024-07-15 15:34:57:000000
191070  2024-07-15 15:34:57:000000  2024-07-15 15:34:57:000000
191071  2024-07-15 15:34:57:000000  2024-07-15 15:34:57:000000
191072  2024-07-15 15:34:57:000000  2024-07-15 15:34:57:000000
191073  2024-07-15 15:34:57:000000  2024-07-15 15:34:57:000000

[191072 rows x 2 columns]

Podemos perceber que nas primeiras e últimas 5 linhas, as colunas **callDate** e **Initiated_at** possuem o mesmo valor. Vamos verificar se esse padrão se repete em todas as linhas da nossa base.

In [12]:
print(f"As colunas callDate e initiatedAt possuem o mesmo valor em {speech[speech['callDate'] == speech['initiatedAt']].shape[0]} linhas")
print(f"As colunas callDate e initiatedAt possuem valores diferentes em {speech[speech['callDate'] != speech['initiatedAt']].shape[0]} linhas")

As colunas callDate e initiatedAt possuem o mesmo valor em 191064 linhas
As colunas callDate e initiatedAt possuem valores diferentes em 8 linhas


In [13]:
speech[speech['callDate'] != speech['initiatedAt']]

audio_length        callId                    callDate  \
587     955.469971  1.546902e+11                         NaN   
692     358.656006  1.546848e+10  2024-07-06 10:21:54:000000   
695     358.656006  1.546848e+10  2024-07-06 10:21:54:000000   
1123          1083  1.546855e+10  2024-07-07 10:36:36:000000   
1182     391.48999  1.546824e+11  2024-07-04 18:27:31:000000   
2728    821.583008  1.546962e+11  2024-07-18 14:20:47:000000   
2919   1211.495972  1.546812e+11                    pos puro   
78864          332  1.546893e+11  2024-07-11 14:57:24:000000   

                      initiatedAt   segment                terminatedAt  \
587    2024-07-12 14:19:12:000000  pos puro  2024-07-12 14:35:07:470000   
692                           NaN  pos puro  2024-07-06 10:27:52:656000   
695                           NaN  pos puro  2024-07-06 10:27:52:656000   
1123                          NaN  pos puro  2024-07-07 10:54:39:000000   
1182                          NaN  pos puro  2024-07-04 18:34:02:490000   
2728                     pos puro  pos puro  2024-07-18 14:34:28:583000   
2919   2024-07-03 12:39:03:000000  pos puro  2024-07-03 12:59:14:496000   
78864  2024-07-ll l4:57:24:000000  pos puro  2024-07-11 15:02:56:000000   

                                reason reason_type  \
587               Informao sobre Plano      speech   
692              Script de Inicializao      script   
695               Questionar Algo Mais      script   
1123   Transferir Pesquisa de Satisfao      script   
1182          2a Via / Cdigo de Barras      speech   
2728                      Cancelamento      speech   
2919               Parcelamento Fatura      speech   
78864              Script de Finalizao      script   

                           subreasons_description  position sgl_uf  customer  \
587          Saber valor pago no plano contratado      10.0     SP     475.0   
692                         Script de Inicializao       0.0     AL     625.0   
695                          Questionar Algo Mais      11.0     AL     625.0   
1123   Transferir ligao para pesquisa de satisfao      10.0     SP     879.0   
1182              Solicitar segunda via de fatura       4.0     DF     896.0   
2728                       Solicitar cancelamento       5.0     SP    1956.0   
2919             Solicitar parcelamento da fatura       9.0     SP    2182.0   
78864                         Script de Finalizao       5.0     SC   11820.0   

         site plan_name  
587    Site 1   Plano 1  
692    Site 1   Plano 1  
695    Site 1   Plano 1  
1123   Site 1   Plano 1  
1182   Site 1   Plano 1  
2728   Site 1   Plano 1  
2919   Site 1   Plano 1  
78864  Site 1   Plano 4

Analisando as 8 linhas em que as colunas possuem valores diferentes, podemos perceber que isso ocorre por causa de valores nulos ou porque a coluna **callDate** recebeu erroneamente o valor da coluna **segment** ("pos puro").

Dessa forma, podemos verificar o padrão de que as colunas **callDate** e **initiatedAt** possuem o mesmo valor, visto que a data em que a ligação ocorreu é a mesma data/hora que o início da ligação. Sendo assim, irei modificar essas linhas para que os valores dessas duas colunas sejam os mesmos.


In [14]:
speech['callDate'] = speech['callDate'].fillna(speech['initiatedAt'])
speech['initiatedAt'] = speech['initiatedAt'].fillna(speech['callDate'])
speech.loc[speech['initiatedAt'] == 'pos puro', "initiatedAt"] = speech['callDate']
speech.loc[speech['callDate'] == 'pos puro', "callDate"] = speech['initiatedAt']

A seguir, vamos analisar se as colunas da base possuem irregularidades utilizando expressões regulares.

A coluna **audio_length** recebe apenas valores numéricos, portanto vou verificar se a coluna possui valores de tipos diferentes disso.

In [15]:
speech[speech['audio_length'].str.contains("[a-zA-Z]", na=False)]

audio_length        callId                    callDate  \
7     29 minutos  1.546791e+11  2024-07-01 11:32:00:000000   
15      55 horas  1.546791e+11  2024-07-01 11:32:00:000000   
249  39 segundos  1.546804e+11  2024-07-02 16:26:44:000000   

                    initiatedAt   segment                terminatedAt  \
7    2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
15   2024-07-01 11:32:00:000000  pos puro  2024-07-01 11:38:45:292000   
249  2024-07-02 16:26:44:000000  pos puro  2024-07-02 16:36:44:387000   

                              reason reason_type  \
7                 Contestao de Conta      speech   
15   Transferir Pesquisa de Satisfao      script   
249             Informao sobre Plano      speech   

                         subreasons_description  position sgl_uf  customer  \
7                        Contestar valor fatura      11.0     SP       1.0   
15   Transferir ligao para pesquisa de satisfao      20.0     SP       1.0   
249        Saber valor pago no plano contratado       5.0     SP     224.0   

       site plan_name  
7    Site 1   Plano 1  
15   Site 1   Plano 1  
249  Site 1   Plano 1

Existem 3 linhas em que a coluna **audio_length** possui caracteres alfabéticos. Como eu não sei a unidade de tempo utilizada nesta coluna, irei remover os caracteres alfabéticos, deixando apenas os númericos e sempre que possível, irei evitar apagar linhas para não perder informação. Após isso, converterei o tipo desta coluna de object para float.

In [16]:
speech['audio_length'] = speech['audio_length'].astype(str).str.replace("[a-zA-Z]", "", regex=True)
speech['audio_length'] = speech['audio_length'].astype(float)

A coluna **callID** possui o identificador único de cada ligação. Tal qual a coluna anterior, vou verificar se ela possui algum valor alfabético.

In [17]:
speech[speech['callId'].astype(str).str.contains("[a-zA-Z]", na=False)]

Empty DataFrame
Columns: [audio_length, callId, callDate, initiatedAt, segment, terminatedAt, reason, reason_type, subreasons_description, position, sgl_uf, customer, site, plan_name]
Index: []

Ela não possui nenhuma irregularidade. Portanto, irei observar a próxima coluna: **callDate**.

In [18]:
speech[~speech['callDate'].astype(str).str.contains("^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}:\d{6}$", na=False)]

Empty DataFrame
Columns: [audio_length, callId, callDate, initiatedAt, segment, terminatedAt, reason, reason_type, subreasons_description, position, sgl_uf, customer, site, plan_name]
Index: []

A coluna **callDate** não possui irregularidades. Sendo assim, irei apenas converter seu tipo para datetime.

In [19]:
speech['callDate'] = pd.to_datetime(speech['callDate'], format='%Y-%m-%d %H:%M:%S:%f')

A seguir, irei aplicar a mesma verificação que fiz anteriormente, porém agora com a coluna **initiatedAt**.

In [20]:
speech[~speech['initiatedAt'].astype(str).str.contains("^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}:\d{6}$", na=False)]

audio_length        callId            callDate  \
78864         332.0  1.546893e+11 2024-07-11 14:57:24   

                      initiatedAt   segment                terminatedAt  \
78864  2024-07-ll l4:57:24:000000  pos puro  2024-07-11 15:02:56:000000   

                    reason reason_type subreasons_description  position  \
78864  Script de Finalizao      script    Script de Finalizao       5.0   

      sgl_uf  customer    site plan_name  
78864     SC   11820.0  Site 1   Plano 4

Existe uma observação na qual o número '1' foi erroneamente trocado por 'l' na coluna **initiatedAt**. Vou corrigir este erro e converterei a coluna para o formato datetime.

In [21]:
speech['initiatedAt'] = speech['initiatedAt'].astype(str).str.replace("l", "1")
speech['initiatedAt'] = pd.to_datetime(speech['initiatedAt'], format='%Y-%m-%d %H:%M:%S:%f')

Vou observar os valores únicos que a coluna **segment** possui.

In [22]:
speech['segment'].unique()

array(['pos puro', 'padaria', 'pos controle', 'pre consumer',
       'pos corporate'], dtype=object)

Podemos perceber que 'padaria' não é um nome valido para representar um grupo de classificação de plano. Por este motivo, irei excluir os registros que possuem o segmento com nome igual a 'padaria'.

In [23]:
speech = speech[speech['segment'] != 'padaria']
speech['segment'] = speech['segment'].astype('string')

Vou verificar se a coluna **terminatedAt** possui algum valor diferente do padrão: 

In [24]:
speech[~speech['terminatedAt'].astype(str).str.contains("^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}:\d{6}$", na=False)]

Empty DataFrame
Columns: [audio_length, callId, callDate, initiatedAt, segment, terminatedAt, reason, reason_type, subreasons_description, position, sgl_uf, customer, site, plan_name]
Index: []

A coluna **terminatedAt** não possui valores incorretos, por isso irei apenas converter seu tipo para datetime.

In [25]:
speech['terminatedAt'] = pd.to_datetime(speech['terminatedAt'], format='%Y-%m-%d %H:%M:%S:%f')

Vou observar os valores únicos que a coluna **reason** possui.

In [26]:
speech['reason'].unique()

array(['Cancelamento', 'Script de Finalizao',
       'Transferir Pesquisa de Satisfao', 'Bloqueio',
       'Contestao de Conta', 'Confirmao Dados', 'Questionar Algo Mais',
       'Script de Inicializao', 'Informar Protocolo',
       'Informao sobre Plano', 'Informao sobre Ofertas',
       '2a Via / Cdigo de Barras',
       'Transferir Pesquisa de Satisfao e Script de Finalizao',
       'Pagamento em Duplicidade', 'Problema Tcnico', 'Mudana de Plano',
       'Aplicativo: Instruo de uso do aplicativo', 'Desconto',
       'Ativao de Servio', 'Demora Atendimento', 'Recarga',
       'Reforo de Sinal', 'Portabilidade', 'Multa Fidelizao',
       'Retorno de Contato', 'Bnus', 'Alterar Data de Vencimento',
       'Declarao Perda e Roubo', 'Alterar Dados Cadastrais',
       'Esforo para Resolver', 'Cliente Rechamante', 'Religa em Confiana',
       'Dependente', 'Informao Pix', 'Induzir ligar novamente',
       'Informar Prazo', 'Dbito Automtico', 'Lentido no Sistema',
       'Esforo Fsico', 'Rec

O padrão que consigo observar é que a coluna **reason** sempre possui a primeira letra iniciando com letra maiuscula ou número. Por isso, vou verificar se ela possui algum resultado iniciando por letra minuscula.

In [27]:
speech[speech['reason'].astype(str).str.contains("^[a-z]", na=False)]

Empty DataFrame
Columns: [audio_length, callId, callDate, initiatedAt, segment, terminatedAt, reason, reason_type, subreasons_description, position, sgl_uf, customer, site, plan_name]
Index: []

Como pode-se observar, esta coluna não possui irregularidades. Dessa forma, vou converter seu tipo para string e analisarei a próxima coluna: **reason_type**

In [28]:
speech['reason'] = speech['reason'].astype('string')

Vamos observar os valores únicos presentes na coluna **reason_type**.

In [29]:
speech['reason_type'].unique()

array(['speech', 'script', 'outros', '-'], dtype=object)

Podemos observar que esta coluna possuí 4 valores únicos: 'speech', 'script', 'outros' e '-'. Os dois primeiros representam respectivamente a fala do cliente e a fala da URA. O valor 'outros' representa outros valores além de 'speech' e 'script'.Como o valor '-' não possui significado claro, irei converte-lo para o valor 'outros', pois ele não corresponde aos valores de 'speech' ou 'script'.

In [30]:
speech['reason_type'] = speech['reason_type'].astype(str).str.replace("-", "outros")
speech['reason_type'] = speech['reason_type'].astype('string')


Assim como a coluna **reason**, a coluna **subreasons_description** apresenta um padrão de ter seus valores iniciando com a primeira letra maiúsucula ou numérica. Dessa forma, irei aplicar a mesma verificação da coluna **reason** para a coluna **subreasons_description**, verificando se ela possui algum valor que se inicia por letra minuscula.

In [31]:
speech[speech['subreasons_description'].astype(str).str.contains("^[a-z]", na=False)]

audio_length        callId            callDate         initiatedAt  \
4041      613.328003  1.546863e+11 2024-07-08 14:24:29 2024-07-08 14:24:29   
114371    342.920013  1.546802e+11 2024-07-02 12:51:31 2024-07-02 12:51:31   
114383    734.909973  1.546893e+11 2024-07-11 14:49:25 2024-07-11 14:49:25   

         segment            terminatedAt              reason reason_type  \
4041    pos puro 2024-07-08 14:34:42.328     Confirmao Dados      script   
114371  pos puro 2024-07-02 12:57:13.920     Confirmao Dados      script   
114383  pos puro 2024-07-11 15:01:39.910  Contestao de Conta      speech   

       subreasons_description  position sgl_uf  customer    site plan_name  
4041                 pos puro       3.0     RJ    3229.0  Site 1   Plano 1  
114371               pos puro       3.0     SP   11871.0  Site 3   Plano 5  
114383               pos puro       4.0     SP   11883.0  Site 3   Plano 5

Podemos verificar que 3 linhas da base possuem a coluna **subreasons_description** com o valor de 'pos puro'. Este valor corresponde ao segmento do plano e não ao detalhamento adicional da reason. Por esse motivo, irei excluir esses registros da base.

In [32]:
speech = speech[speech['subreasons_description'] != 'pos puro']
speech['subreasons_description'] = speech['subreasons_description'].astype('string')

A coluna **position** representa a ordenação numérica da fala ao longo da conversa. Dessa maneira, vou verificar se ela possui algum valor alfabético.

In [33]:
speech[speech['position'].astype(str).str.contains("[a-zA-Z]", na=False)]

Empty DataFrame
Columns: [audio_length, callId, callDate, initiatedAt, segment, terminatedAt, reason, reason_type, subreasons_description, position, sgl_uf, customer, site, plan_name]
Index: []

Esta coluna não possui irregularidades. Sendo assim, vou observar a coluna **sgl_uf**. Primeiro irei verificar se ela possui algum valor diferente do padrão de representação de uma unidade federativa e após isso, irei confirmar a validade da coluna, verificando se existem no máximo 27 valores únicos, visto que no Brasil existem 27 unidades federativas.

In [34]:
speech[~speech['sgl_uf'].astype(str).str.contains("^[A-Z]{2}$", na=False)]

audio_length        callId            callDate         initiatedAt  \
3259     605.840027  1.546804e+11 2024-07-02 16:25:13 2024-07-02 16:25:13   
78368    576.000000  1.546813e+11 2024-07-03 15:25:31 2024-07-03 15:25:31   

        segment            terminatedAt  \
3259   pos puro 2024-07-02 16:35:18.840   
78368  pos puro 2024-07-03 15:35:07.000   

                                         reason reason_type  \
3259                       Informao sobre Plano      speech   
78368  Aplicativo: Instruo de uso do aplicativo      script   

                           subreasons_description  position      sgl_uf  \
3259      Saber detalhes sobre limite de internet      11.0    pos puro   
78368  Instruo de uso de aplicativos corporativos       1.0  Terra mdia   

       customer    site plan_name  
3259     2594.0  Site 1   Plano 1  
78368   11162.0  Site 1   Plano 4

A coluna **sgl_uf** possui 2 valores que não correspondem ao padrão de unidade federativa: 'pos puro' e 'Terra mdia'. Portanto, irei apagar as linhas da base que contém estes registros e irei verificar se a coluna possui no máximo 27 valores correspondendo às unidades federativas do Brasil.

In [35]:
speech = speech[~speech['sgl_uf'].isin(['pos puro', 'Terra mdia'])]
speech['sgl_uf'] = speech['sgl_uf'].astype('string')
speech['sgl_uf'].unique()

<StringArray>
['SP', 'MG', 'PR', 'RJ', 'PE', 'SC', 'TO', 'RS', 'PA', 'BA', 'SE', 'AL', 'AM',
 'CE', 'DF', 'GO', 'MA', 'ES', 'PB', 'MT', 'RN', 'MS', 'PI', 'AP', 'AC', 'RR',
 'RO']
Length: 27, dtype: string

Após o tratamento, podemos perceber que a coluna **sgl_uf** possui 27 valores únicos que correspondem às unidades federativas do Brasil. Sendo assim, podemos tratar a próxima coluna: **customer**.

A coluna **customer** representa o identificador númerico do cliente. Por isso, vou verificar se ela possui algum valor alfabético.

In [36]:
speech[speech['customer'].astype(str).str.contains("[a-zA-Z]", na=False)]


Empty DataFrame
Columns: [audio_length, callId, callDate, initiatedAt, segment, terminatedAt, reason, reason_type, subreasons_description, position, sgl_uf, customer, site, plan_name]
Index: []

Não existem irregularidades na coluna **customer**. Vamos então ver os valores únicos presentes na coluna **site**.

In [37]:
speech['site'].unique()

array(['Site 1', 'Venha ver a promoo do milho', 'pos puro', 'Site 2',
       'Site 3', 'Site 4', 'Site 5', 'Site 6'], dtype=object)

Podemos perceber que 2 valores estão fora do padrão desta coluna: 'Venha ver a promoo do milho' e 'pos puro'. Desse modo, vamos eliminar as linhas que contém estes registros da base.

In [38]:
speech = speech[~speech['site'].isin(['Venha ver a promoo do milho', 'pos puro'])]
speech['site'] = speech['site'].astype('string')

Por fim, vamos tratar a última coluna da base: a coluna **plan_name**. Vamos verificar os valores únicos presentes nela: 

In [39]:
speech['plan_name'].unique()

array(['Plano 1', 'pos puro', 'Plano 2', 'Plano 3', 'Plano 4', 'Plano 5',
       'Plano 6', 'Plano 7', 'Plano 8', 'Plano 9', 'Plano 10', 'Plano 11',
       'Plano 12', 'Plano 13', 'Plano 14', 'Plano 15', 'Plano 16',
       'Plano 17', 'Plano 18', 'Plano 19', 'Plano 20', 'Plano 21',
       'Plano 22', 'Plano 23', 'Plano 24', 'Plano 25', 'Plano 26',
       'Plano 27', 'Plano 28', 'Plano 29', 'Plano 30', 'Plano 31',
       'Plano 32', 'Plano 33', 'Plano 34', 'Plano 35', 'Plano 36',
       'Plano 37', 'Plano 38', 'Plano 39', 'Plano 40', 'Plano 41',
       'Plano 42', 'Plano 43', 'Plano 44', 'Plano 45', 'Plano 46',
       'Plano 47', 'Plano 48', 'Plano 49', 'Plano 50'], dtype=object)

Observando atentamente, percebe-se que esta coluna possuí um valor fora do padrão: **'pos puro'**. Sendo assim, vou excluir as linhas que contém este valor da base.

In [40]:
speech = speech[speech['plan_name'] != 'pos puro']
speech['plan_name'] = speech['plan_name'].astype('string')

In [41]:
speech.info()

<class 'pandas.core.frame.DataFrame'>
Index: 191062 entries, 0 to 191073
Data columns (total 14 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   audio_length            191062 non-null  float64       
 1   callId                  191062 non-null  float64       
 2   callDate                191062 non-null  datetime64[ns]
 3   initiatedAt             191062 non-null  datetime64[ns]
 4   segment                 191062 non-null  string        
 5   terminatedAt            191062 non-null  datetime64[ns]
 6   reason                  191062 non-null  string        
 7   reason_type             191062 non-null  string        
 8   subreasons_description  191062 non-null  string        
 9   position                191062 non-null  float64       
 10  sgl_uf                  191062 non-null  string        
 11  customer                191062 non-null  float64       
 12  site                    191062 non-

A base foi tratada com sucesso e todos os campos estão com o tipo correto. Irei importa-la para o formato csv e prosseguirei para a próxima etapa do case: *Geração de Análises*.

In [62]:
speech.to_csv('base_tratada.csv', sep=';', encoding='utf-8-sig', index=False)